NeoML/include/NeoML/Dnn/Dnn.h

```diff
@@ -646,8 +655,45 @@ class NEOML_API CDnn : public CDnnLayerGraph {
 	friend class CBaseLayer;
 	friend class CCompositeLayer;
 	friend class CRecurrentLayer;
+	friend class CDnnRefManager;
+};
+
+//------------------------------------------------------------------------------------------------------------
+
+// Manages reference counts and initial states for CDnn objects, restoring configurations once all references are removed
+class NEOML_API CDnnRefManager : public IObject {
+public:
+	CDnnRefManager(CDnn* _originalDnn) :
+		learningState(_originalDnn->IsLearningEnabled())
+	{
+		sharedDnns.Add(_originalDnn);
+	}
+
+private:
+	bool learningState; // initial isLearningEnable state before creating reference copies
+	CHashTable<CDnn*> sharedDnns; // set of dnns with shared weights
```

Comment:
Why is just `int` counter inside is not enough?
`counter = -1` --> is CReferenceDnn, never learn
`counter = 0` --> is normal Dnn
`counter > 0` --> Dnn has CReferenceDnns, disable learning

Reply:
We will need to have pointer to dnn that we consider original to restore it
I think it's more optimal for all references no to depend on first original dnn since they are all actually equal
For example, if there's parent and child dnns, after removing parent, child will take its state

Reply:
Are you sure, that original dnn and reference dnn are equal?
The types of the blobs in the reference dnn are deffernt type and has the pointers to parents

My idea will work if all dnns store the pointers to the unique object of `refInfo`

Reply:
yep, my bad, you are right, reference dnn has less functionality (no serialization, creating own references etc)

Reply:
`no serialization`, are there corresponding asserts about all of its rejections?